File: RECETAS-API/api/models/receta.php
<?php
class Receta {
    private $conn;
    private $table_name = "recetas";

    public $id;
    public $titulo;
    public $descripcion;
    public $ingredientes;
    public $instrucciones;
    public $tiempo_preparacion;
    public $porciones;
    public $dificultad;
    public $categoria_id;
    public $imagen_url;
    public $autor;
    public $es_premium;
    public $created_at;

    public function __construct($db) {
        $this->conn = $db;
    }

    // Obtener todas las recetas
    public function read() {
        $query = "SELECT r.*, c.nombre as categoria_nombre 
                  FROM " . $this->table_name . " r 
                  LEFT JOIN categorias c ON r.categoria_id = c.id 
                  ORDER BY r.created_at DESC";
        
        $stmt = $this->conn->prepare($query);
        $stmt->execute();
        return $stmt;
    }

    // Obtener una receta por ID
    public function readOne() {
        $query = "SELECT r.*, c.nombre as categoria_nombre 
                  FROM " . $this->table_name . " r 
                  LEFT JOIN categorias c ON r.categoria_id = c.id 
                  WHERE r.id = ? LIMIT 0,1";
        
        $stmt = $this->conn->prepare($query);
        $stmt->bindParam(1, $this->id);
        $stmt->execute();
        
        $row = $stmt->fetch(PDO::FETCH_ASSOC);
        
        if($row) {
            $this->titulo = $row['titulo'];
            $this->descripcion = $row['descripcion'];
            $this->ingredientes = $row['ingredientes'];
            $this->instrucciones = $row['instrucciones'];
            $this->tiempo_preparacion = $row['tiempo_preparacion'];
            $this->porciones = $row['porciones'];
            $this->dificultad = $row['dificultad'];
            $this->categoria_id = $row['categoria_id'];
            $this->imagen_url = $row['imagen_url'];
            $this->autor = $row['autor'];
            $this->es_premium = $row['es_premium'];
            $this->created_at = $row['created_at'];
            return true;
        }
        return false;
    }

    // Crear nueva receta
    public function create() {
        $query = "INSERT INTO " . $this->table_name . " 
                  SET titulo=:titulo, descripcion=:descripcion, ingredientes=:ingredientes, 
                      instrucciones=:instrucciones, tiempo_preparacion=:tiempo_preparacion, 
                      porciones=:porciones, dificultad=:dificultad, categoria_id=:categoria_id, 
                      imagen_url=:imagen_url, autor=:autor, es_premium=:es_premium";
        
        $stmt = $this->conn->prepare($query);
        
        $stmt->bindParam(":titulo", $this->titulo);
        $stmt->bindParam(":descripcion", $this->descripcion);
        $stmt->bindParam(":ingredientes", $this->ingredientes);
        $stmt->bindParam(":instrucciones", $this->instrucciones);
        $stmt->bindParam(":tiempo_preparacion", $this->tiempo_preparacion);
        $stmt->bindParam(":porciones", $this->porciones);
        $stmt->bindParam(":dificultad", $this->dificultad);
        $stmt->bindParam(":categoria_id", $this->categoria_id);
        $stmt->bindParam(":imagen_url", $this->imagen_url);
        $stmt->bindParam(":autor", $this->autor);
        $stmt->bindParam(":es_premium", $this->es_premium);
        
        if($stmt->execute()) {
            return true;
        }
        return false;
    }

    // Buscar recetas
    public function search($keywords) {
        $query = "SELECT r.*, c.nombre as categoria_nombre 
                  FROM " . $this->table_name . " r 
                  LEFT JOIN categorias c ON r.categoria_id = c.id 
                  WHERE r.titulo LIKE ? OR r.descripcion LIKE ? OR r.ingredientes LIKE ?
                  ORDER BY r.created_at DESC";
        
        $stmt = $this->conn->prepare($query);
        $keywords = "%{$keywords}%";
        $stmt->bindParam(1, $keywords);
        $stmt->bindParam(2, $keywords);
        $stmt->bindParam(3, $keywords);
        $stmt->execute();
        
        return $stmt;
    }
}
?>

File: api/models/receta.py
# -*- coding: utf-8 -*-
"""
Define los modelos de datos para las recetas.

Este archivo contiene:
- `RecetaBase`: Un modelo Pydantic que define los campos base para una receta.
- `RecetaCreate`: Un modelo Pydantic para la creación de una nueva receta, con validaciones.
- `Receta`: Un modelo Pydantic que representa una receta completa, incluyendo campos adicionales.
- `RecetaDB`: Un modelo SQLAlchemy que define la tabla `recetas` en la base de datos.
"""

from pydantic import BaseModel, Field
from typing import Optional
from datetime import datetime
from sqlalchemy import Column, Integer, String, Boolean, DateTime, ForeignKey, func
from sqlalchemy.orm import relationship
from sqlalchemy.dialects.mssql import NVARCHAR
from ..database import Base

# -------------------
# Modelos Pydantic
# -------------------

class RecetaBase(BaseModel):
    """
    Modelo Pydantic base para una receta.

    Define los campos comunes que se utilizan para la creación y representación de una receta.
    """
    titulo: str
    descripcion: Optional[str] = None
    ingredientes: str
    instrucciones: str
    dificultad: Optional[str] = 'Fácil'
    categoria_id: int
    imagen_url: Optional[str] = None
    autor: Optional[str] = 'Anónimo'
    es_premium: Optional[bool] = False
    activo: Optional[bool] = True
    updated_at: Optional[datetime] = None

class RecetaCreate(RecetaBase):
    """
    Modelo Pydantic para la creación de una nueva receta.

    Hereda de `RecetaBase` y añade validaciones para `tiempo_preparacion` y `porciones`.
    Se utiliza para validar los datos de entrada al crear una nueva receta.
    `Field` se usa para añadir validaciones a los campos.
    """
    tiempo_preparacion: int = Field(..., gt=0, description="Tiempo de preparación en minutos, debe ser mayor a 0.")
    porciones: int = Field(..., gt=0, le=20, description="Número de porciones, debe ser mayor a 0 y menor o igual a 20.")

class Receta(RecetaBase):
    """
    Modelo Pydantic para representar una receta completa.

    Hereda de `RecetaBase` y añade campos que son generados por la base de datos
    o que se obtienen de otras tablas (como `categoria_nombre`).
    """
    id: int
    categoria_nombre: Optional[str] = None  # Este campo se poblará con un JOIN
    vistas: int
    likes: int
    created_at: datetime

    class Config:
        """
        Configuración del modelo Pydantic.

        `from_attributes = True` permite que el modelo Pydantic se cree a partir de
        un objeto SQLAlchemy.
        """
        from_attributes = True

# -------------------
# Modelo SQLAlchemy
# -------------------

class RecetaDB(Base):
    """
    Modelo SQLAlchemy para la tabla `recetas`.

    Define la estructura de la tabla `recetas` en la base de datos.
    """
    __tablename__ = "recetas"

    id = Column(Integer, primary_key=True, index=True)
    titulo = Column(String(255), nullable=False)
    descripcion = Column(NVARCHAR('max'), nullable=True)
    ingredientes = Column(NVARCHAR('max'), nullable=False)
    instrucciones = Column(NVARCHAR('max'), nullable=False)
    tiempo_preparacion = Column(Integer, default=0)
    porciones = Column(Integer, default=1)
    dificultad = Column(String(50), default='Fácil')
    categoria_id = Column(Integer, ForeignKey("categorias.id"), nullable=False)
    imagen_url = Column(String(500), nullable=True)
    autor = Column(String(100), default='Anónimo')
    es_premium = Column(Boolean, default=False)
    activo = Column(Boolean, default=True)
    vistas = Column(Integer, default=0)
    likes = Column(Integer, default=0)
    created_at = Column(DateTime, server_default=func.now())
    updated_at = Column(DateTime, onupdate=func.now())

    # Define la relación con la tabla `categorias`.
    # Esto permite acceder a la categoría de una receta como un objeto.
    categoria = relationship("CategoriaDB")